### DISCUSSION TITLE
Add a way to preload/preconnect injected script/css files

### DISCUSSION THREAD

**Discussion by jagu-sayan** [withastro/roadmap#609] [Proposal]
### Body

# Summary

I want a simple way to `preload` or `preconnect` injected script/css/fonts files.

# Background & Motivation

It's help to improve performance and web vitals, see:
  - https://portal.gitnation.org/contents/top-core-web-vitals-recommendations-for-2023
  - https://portal.gitnation.org/contents/http3-performance-for-js-developers 
It's impossible to customize injected script and know path names in advance.

Exemple:
```html
Actual injected script in my html page:
<link href="/_astro/404.ae8dd206.css" rel="stylesheet">
<script src="/_astro/hoisted.bff9a731.js" type="module"></script>
I want:
<link rel="preload" as="style" href="/_astro/404.ae8dd206.css" />
<link rel="preload" as="script" href="/_astro/hoisted.bff9a731.js" />
<link href="/_astro/404.ae8dd206.css" rel="stylesheet">
<script src="/_astro/hoisted.bff9a731.js" type="module"></script>
```

# Goals

- Speed up improvement and better web vitals score
- Allow to enable [Early Hints on Cloudflare page](https://blog.cloudflare.com/fr-fr/early-hints-on-cloudflare-pages-fr-fr/)
- Allow more flexibility for future usage or usage I didn't know

# Proposal

Add an `Astro.injectedScripts` in Runtime API:
```typescript
type InjectedScript = {
  path: string;
  as: string;
};

Astro.injectedScripts: InjectedScript [] = [{
  path: "/_astro/404.ae8dd206.css",
  as: "style",
}, {
  path: "/_astro/hoisted.bff9a731.js",
  as: "script",
}];
```

Layout.astro:
```astro
<html>
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width" />
    <title>My webpage</title>
    {Astro.injectedScripts(({path, as}) => (
      <link rel="preload" as={as} href={path} />
    ))}
  </head>
  <body>
     <h1>42</h1>
  </body>
</html>
```

Limitation:
- only useful for production build
- my proposal is not very flexible, suggestions ?

Related to https://github.com/withastro/astro/issues/5302
I try this vite plugin without success : https://github.com/Applelo/vite-plugin-inject-preload

**Comment by natemoo-re** [comment]
Related proposal https://github.com/withastro/roadmap/discussions/561

**Comment by wtchnm** [comment]
As an alternative, here is a [middleware](https://docs.astro.build/en/guides/middleware/) that add preloads for CSS files using [rehype](https://github.com/rehypejs/rehype):

```typescript
import { defineMiddleware } from 'astro:middleware'
import type { Element, Root } from 'hast'
import { rehype } from 'rehype'
import { visit } from 'unist-util-visit'
import { isCssLink } from 'hast-util-is-css-link'

const rehypeInstance = rehype().use(() => (tree: Root) => {
	let head: Element | undefined
	const links: Element[] = []
	visit(tree, 'element', node => {
		if (node.tagName === 'head') head = node
		if (isCssLink(node))
			links.push({
				type: 'element',
				tagName: 'link',
				properties: { rel: 'preload', as: 'style', href: node.properties.href },
				children: []
			})
	})
	head?.children.unshift(...links)
})

export const onRequest = defineMiddleware(async (_, next) => {
	const response = await next()
	if (response.headers.get('content-type') !== 'text/html') return response

	const html = await response.text()
	const file = await rehypeInstance.process(html)

	return new Response(file.toString(), response)
})

```

**Comment by josephmaxim** [comment]
+1 the bundled css is render blocking so this should be high priority.

**Comment by aldo-roman** [comment]
For anyone wondering, there is a way to achieve this. See https://docs.astro.build/en/guides/styling/#url-css-imports

```
---
import stylesUrl from '../styles/main.css?url';
---
<link rel="preload" href={stylesUrl} as="style">
```

**Comment by cerpow** [comment]
If you're using Sass, hot reload stops working. Here's what I did to fix it:

```
---
import '../css/app.scss'; 
import stylesUrl from '../css/app.scss?url';
---

<link rel="preload" href={stylesUrl} as="style" />
```

During the build step, Astro replaces the URLs with the generated .css file.

**Comment by delucis** [comment]
Just noting that for most generic use cases, this is _not necessary_. For a browser, there is no difference between:

```html
<head>
  <link href="/_astro/404.ae8dd206.css" rel="stylesheet">
</head>
```

and

```html
<head>
  <link rel="preload" as="style" href="/_astro/404.ae8dd206.css" />
  <link href="/_astro/404.ae8dd206.css" rel="stylesheet">
</head>
```

With or without a `<link rel="preload">`, the browser’s preload scanner will discover scripts and styles in the `<head>` and load them the same way.

In general, you only need a preload if the resource is _not_ also included in the head, e.g. because it’s something loaded dynamically (via a script for example).

It sounds like doing this might still help in the Cloudflare Pages “Early Hints” use case as it appears that feature parses preload tags and use them for that, but otherwise it generally shouldn’t be needed.

As [mentioned above](https://github.com/withastro/roadmap/discussions/609#discussioncomment-12122912), when you do want to do this you can use a `?url` import:

```astro
---
import scriptUrl from '../some-script.ts?url';
import styleUrl from '../some-style.css?url';
---

<link rel="preload" href={scriptUrl} as="script">
<link rel="preload" href={styleUrl} as="style">
```

**Comment by palockocz** [comment]
Sorry, I deleted my question because I found solution just right after I posted it.

Your example doesn't work for me. I must use ?worker&url to get working URL to TS file.
(I need to preload the script because of PageSpeed says that the script is blocking requests.)

But another problem is, that the builded JS file isn't minified. Is some param to force minified version?

// EDIT: I have already solved the file minification, using rollupOptions

**Reply by delucis** [reply]
> I need to preload the script because of PageSpeed says that the script is blocking requests.

> the builded JS file isn't minified.

How do you author your script? An Astro-bundled `<script>` should not be render blocking and should be minified, so might need to see your code to understand what you’re doing.

**Reply by palockocz** [reply]
I had SwiperJS script inside of <script> tag with some TypeScript in index.astro which was builded by Astro to a separate JS file.

Then I need to preload it, so I moved it from <script> tag to separate TS file, import using ?worker&url and solved the minification using rollupOptions inside of astro.config.mjs.

**Comment by cerpow** [comment]
Yes, if you don't use is:inline, Astro should handle it properly
